Web API 8 Actions and functions (#263)
MicrosoftDocs/OData-docs: Odata-docs/webapi-8/fundamentals/actions-functions.md

---
title: "Actions and functions"
description: "This documentation describes how to add actions and functions."

author: KenitoInc
ms.author: kemunga

ms.date: 9/15/2022
---
# Actions and functions
**Applies To**:[!INCLUDE[appliesto-webapi](../../includes/appliesto-webapi-v8.md)]

> In OData, actions and functions are a way to add server-side behaviors that are not easily defined as CRUD operations on entities. This documentation shows how to add actions and functions to an OData v4 endpoint, using Web API v8.x.

Reviewer: web api v 8.x ==> ASP.NET Core OData v8.x

> In OData, actions and functions are a way to add server-side behaviors that are not easily defined as CRUD operations on entities. This documentation shows how to add actions and functions to an OData v4 endpoint, using Web API v8.x.

Reviewer: CRUD (aks. ....)

> In OData, actions and functions are a way to add server-side behaviors that are not easily defined as CRUD operations on entities. This documentation shows how to add actions and functions to an OData v4 endpoint, using Web API v8.x.

Reviewer 2: ```suggestion
In OData, actions and functions are a way to add server-side behavior that are not easily defined as CRUD operations on entities. This documentation shows how to add actions and functions to an OData v4 endpoint, using Web API v8.x.
```

## What are actions and functions?
- Functions are operations exposed by an OData service that MUST return data and MUST have no observable side effects. They may support further composition. Functions are invoked by using HTTP GET requests.
- Actions are operations exposed by an OData service that MAY have side effects when invoked. Actions have a side effect on the server, so they are invoked by using HTTP POST requests. They cannot be further composed in order to avoid non-deterministic behavior.

## Difference between bound and unbound operations.
From OData V4 spec, functions and actions can be either bound to a type or unbound.
Bound operations are bound to an entity type, primitive type, complex type, or a collection.
Unbound operations are invoked as static operations on the service.

## Practical examples
We will add detailed examples on how to configure bound and unbound functions and actions.
We will start by setting up the OData WebAPI 8.x project.

> ### CLR classes

Reviewer: CLR ==> Data Model classes

{
    public string ID { get; set; }
    public int Rating { get; set; }
    public string BookID { get; set; }
}
```

### Controller
```csharp
public class BooksController : ODataController
{
    // Get ~/Books
    [EnableQuery]
    public IActionResult Get()
    {
        return Ok(DataSource.Instance.Books);
    }
}
```

### Build Edm Model
```csharp
private static IEdmModel GetEdmModel()
{
    var builder = new ODataConventionModelBuilder();
    builder.EntitySet<Book>("books");
    var model = builder.GetEdmModel();
    return model;
}
```

## Bound Function
Bound functions can be bound to an entity type or a collection of entity type.
In our example model above, the bound function can be bound to a single book or a collection of books.
We will demonstrated how to add a function that is bound to a collection.

In the Edm Model, we modify the code as follows.

```csharp
private static IEdmModel GetEdmModel()
{
    var builder = new ODataConventionModelBuilder();
    builder.EntitySet<Book>("books");
    builder.EntityType<Book>().Collection
        .Function("mostRecent")
        .Returns<string>();
    var model = builder.GetEdmModel();
    return model;
}
```

> In the Edm Model above, we are adding a `mostRecent` function bound to the books collection. The function returns the `ID` of the most recent book. The `ID` is of `string` type.

Reviewer 3: suggestion: Mention that operations are case insensitive.

Author: I deliberately ignored this since I will need to show how to configure case insensitivity. Also in these sample we are using attribute routing instead of convention routing

We will update the `BooksController` as follows.

```csharp
public class BooksController : ODataController
{
    // Get ~/Books
    [EnableQuery]
    public IActionResult Get()
    {
        return Ok(DataSource.Instance.Books);
    }

>     [HttpGet("odata/Books/MostRecent()")]

Reviewer 3: ```suggestion
    [HttpGet("odata/Books/mostRecent()")]
```

    public IActionResult MostRecent()
    {
        var maxBookId = DataSource.Instance.Books.Max(x => x.ID);
        return Ok(maxBookId);
    }
}
```

If we invoke `GET odata/Books/MostRecent()`
We get the response below:

```json
{
    "@odata.context": "http://localhost:5000/odata/$metadata#Edm.String",
    "value": "8"
}
```

## Unbound function

> Unbound functions don’t bound to any type and they are called as static operations. All unbound function overloads MUST have same return type.

Reviewer: "they are called as static operations"  => this could be confusing. We can say " they look like a static method in C#" ?

In the Edm Model, we modify the code as follows.

```csharp
private static IEdmModel GetEdmModel()
{
    var builder = new ODataConventionModelBuilder();
    builder.EntitySet<Book>("books");
    builder.EntityType<Book>().Collection
        .Function("mostRecent")
        .Returns<string>();
    builder.Function("returnAllForKidsBooks").ReturnsFromEntitySet<Book>("books");
    var model = builder.GetEdmModel();
    return model;
}
```

In the Edm Model above, we are adding a `returnAllForKidsBooks` function. The function returns a collection of books.

We will update the `BooksController` as follows.

```csharp
public class BooksController : ODataController
{
    // Get ~/Books
    [EnableQuery]
    public IActionResult Get()
    {
        return Ok(DataSource.Instance.Books);
    }

    [HttpGet("odata/Books/MostRecent()")]
    public IActionResult MostRecent()
    {
        var maxBookId = DataSource.Instance.Books.Max(x => x.ID);
        return Ok(maxBookId);
    }

>     [HttpGet("odata/ReturnAllForKidsBooks")]

Reviewer: mention: why we should have () for one function, the other without  () ?

            "year": 2004,
            "forKids": true
        },
        {
            "id": "6",
            "isbn": "FF0011",
            "title": "Book 6",
            "year": 2005,
            "forKids": true
        },
        {
            "id": "7",
            "isbn": "GG0011",
            "title": "Book 7",
            "year": 2006,
            "forKids": true
        }
    ]
}
```

## Bound Action
Similar to bound functions, bound actions can be bound to entity type or collection of entity type.
However, the overload of bound actions are different. Bound actions can be overloaded, but overload must happen by different binding parameter. For one binding parameter, there can be only one bound action.

In the Edm Model, we modify the code as follows.

```csharp
private static IEdmModel GetEdmModel()
{
    var builder = new ODataConventionModelBuilder();
    builder.EntitySet<Book>("books");
    builder.EntityType<Book>().Collection
        .Function("mostRecent")
        .Returns<string>();
    builder.Function("returnAllForKidsBooks").ReturnsFromEntitySet<Book>("books");
    builder.EntityType<Book>()
        .Action("rate")
        .Parameter<int>("rating");

>     var model = builder.GetEdmModel();

Reviewer: omit the exsiting codes

We will update the `BooksController` as follows.

```csharp
public class BooksController : ODataController
{
    // Get ~/Books
    [EnableQuery]
    public IActionResult Get()
    {
        return Ok(DataSource.Instance.Books);
    }

    [HttpGet("odata/Books/MostRecent()")]
    public IActionResult MostRecent()
    {
        var maxBookId = DataSource.Instance.Books.Max(x => x.ID);
        return Ok(maxBookId);
    }

    [HttpGet("odata/ReturnAllForKidsBooks")]
    public IActionResult ReturnAllForKidsBooks()
    {
        var forKidsBooks = DataSource.Instance.Books.Where(m => m.ForKids == true);
        return Ok(forKidsBooks);
    }

    [HttpPost("odata/Books({key})/Rate")]
    public IActionResult Rate([FromODataUri] string key, ODataActionParameters parameters)
    {
        if (!ModelState.IsValid)
        {
            return BadRequest();
        }

        int rating = (int)parameters["rating"];

        if (rating < 0)
        {
            return BadRequest();
        }

        return Ok(new BookRating() { BookID = key, Rating = rating });
    }
}
```

> If we invoke `POST odata/Books('1')/Rate {"rating": 7}`

Reviewer 2: I think we should heave the request URL and body on separate line to make it clear the the JSON part is the body:

```
POST odata/Books('1')/Rate

{"rating": 7}
```

We get the response below:

```json
{
    "id": null,
    "rating": 7,
    "bookID": "1"
}
```

## Unbound Action

> Unbound actions don’t bound to any type and they are called as static operations same as unbound functions. However, unbound actions do not allow overloading.

Reviewer 2: Use "don't bind to any type" or "are not bound to any type"

---

> title: "Actions and functions"

Reviewer 4: ```suggestion
title: "Actions and Functions"
```

description: "This documentation describes how to add actions and functions."

author: KenitoInc
ms.author: kemunga

ms.date: 9/15/2022
---
# Actions and functions
**Applies To**:[!INCLUDE[appliesto-webapi](../../includes/appliesto-webapi-v8.md)]

> In OData, actions and functions are a way to add server-side behavior that are not easily defined as CRUD operations on entities. This documentation shows how to add actions and functions to an OData v4 endpoint, using ASP.NET Core OData v8.x.

Reviewer 5: ```suggestion
In OData, actions and functions are a way to add server-side behavior that is not easily defined as CRUD operations on entities. This documentation shows how to add actions and functions to an OData v4 endpoint, using ASP.NET Core OData v8.x.
```

> In OData, actions and functions are a way to add server-side behavior that are not easily defined as CRUD operations on entities. This documentation shows how to add actions and functions to an OData v4 endpoint, using ASP.NET Core OData v8.x.

Reviewer 2: I think there was a request to explain what CRUD means, e.g., `CRUD operations (Create-Read-Update-Delete)`.

## What are actions and functions?

> - `Functions` are operations exposed by an OData service that MUST return data and MUST have no observable side effects. They may support further composition. Functions are invoked by using HTTP GET requests.

Reviewer 2: I could be wrong, but I think we can invoke functions with `POST` requests, for example when passing parameters in the body if the inputs are complex objects.

- `Actions` are operations exposed by an OData service that MAY have side effects when invoked. Actions have a side effect on the server, so they are invoked by using HTTP POST requests. They cannot be further composed in order to avoid non-deterministic behavior.

> Side effects means that actions results in a change in the server; either a resource is updated, deleted or a new resource is created.

Reviewer 2: ```suggestion
Side effects means that actions results in a change in the server; for example a resource is updated, deleted or a new resource is created.
```

    opt.RouteOptions.EnableNonParenthesisForEmptyParameterFunction = true;
});
```

### Bound Action
Actions are invoked via `POST` requests therefore we can pass a payload in the body.

```json
POST ~/odata/Books('1')/Rate
{
    "rating": 7
}
```

### Unbound action
```json
POST ~/odata/incrementBookYear
{
    "increment": 7,
    "id": "1"
}
```

## Practical examples
We will add detailed examples on how to configure bound and unbound functions and actions.
We will start by setting up the ASP.NET Core OData v8.x project.

### Data Model classes
```csharp
public class Book
{
    public string ID { get; set; }
    public string Title { get; set; }
    public bool ForKids { get; set; }
}

public class BookRating
{
    public string ID { get; set; }
    public int Rating { get; set; }
    public string BookID { get; set; }
}
```

### Controller
```csharp
public class BooksController : ODataController
{
    // Get ~/Books
    [EnableQuery]
    public IActionResult Get()
    {
        return Ok(DataSource.Instance.Books);
    }
}
```

### Build Edm Model
```csharp
private static IEdmModel GetEdmModel()
{

>     var builder = new ODataConventionModelBuilder();

Reviewer 2: `ODataConventionModelBuilder()` is in a separate NuGet package: `Microsoft.OData.ModelBuilder`. Maybe that's worth mentioning? Or maybe that'll be covered in the "Getting Started" guide?

Author: Model builder should not be covered in this section

    builder.EntitySet<Book>("books");
    var model = builder.GetEdmModel();
    return model;
}
```

## Bound Function
Bound functions can be bound to an entity type or a collection of entity type.
In our example model above, the bound function can be bound to a single book or a collection of books.
We will demonstrated how to add a function that is bound to a collection.

> 

Reviewer 2: I think for the sake of being exhaustive, we should have examples for binding both to collections and single entities, with and without parameters, with and without return type. The reasoning is that this should be the authoritative source of the truth, if it's not mentioned here, we shouldn't expect the user to know how to use it or to know whether it exists.

Author: That will be too many examples in one section considering each example has code in the model & controller and request & response examples.

Reviewer 2: I get that. But where else can the user get this information? That user doesn't have to read all the examples, but if they need to do something or to find out something, it would be great if they can find it in the docs. In my opinion, having "too many examples" is a result of OData supporting many different options. I think it's worse to support things we don't demonstrate how to use.

In the Edm Model, we modify the code as follows.

```csharp
private static IEdmModel GetEdmModel()
{
    var builder = new ODataConventionModelBuilder();
    // -------------
    // Existing code
    // -------------

    builder.EntityType<Book>().Collection
        .Function("mostRecent")
        .Returns<string>();
    var model = builder.GetEdmModel();
    return model;
}
```

In the Edm Model above, we are adding a `mostRecent` function bound to the books collection. The function returns the `ID` of the most recent book. The `ID` is of `string` type.

We will update the `BooksController` as follows.

```csharp
public class BooksController : ODataController
{
    // -------------
    // Existing code
    // -------------

    [HttpGet("odata/Books/mostRecent()")]
    public IActionResult MostRecent()
    {
        var maxBookId = DataSource.Instance.Books.Max(x => x.ID);
        return Ok(maxBookId);
    }
}
```

If we invoke `GET odata/Books/mostRecent()`
We get the response below:

```json
{
    "@odata.context": "http://localhost:5000/odata/$metadata#Edm.String",
    "value": "8"
}
```

## Unbound function

> Unbound functions don’t bind to any type and they are invoked as static operations. All unbound function overloads MUST have same return type.

Reviewer 2: I think we should have a separate subsection for function overloads, so we can explain what overloads are, the limitations and provide examples. Same thing for actions.
